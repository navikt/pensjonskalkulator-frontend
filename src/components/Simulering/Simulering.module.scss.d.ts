import globalClassNames from '../../style.d'
declare const classNames: typeof globalClassNames & {
  readonly section: 'section'
  readonly info: 'info'
  readonly buttonRow: 'buttonRow'
  readonly buttonRow__visible: 'buttonRow__visible'
  readonly buttonRowElement: 'buttonRowElement'
  readonly buttonRowElement__Right: 'buttonRowElement__Right'
  readonly loader: 'loader'
  readonly infoIcon: 'infoIcon'
  readonly infoText: 'infoText'
  readonly error: 'error'
  readonly errorIcon: 'errorIcon'
  readonly errorText: 'errorText'
  readonly tooltip: 'tooltip'
  readonly tooltipTable: 'tooltipTable'
  readonly tooltipLine: 'tooltipLine'
  readonly tooltipTableHeaderCell: 'tooltipTableHeaderCell'
  readonly tooltipTableCell: 'tooltipTableCell'
  readonly tooltipTableHeaderCell__left: 'tooltipTableHeaderCell__left'
  readonly tooltipTableHeaderCell__right: 'tooltipTableHeaderCell__right'
  readonly tooltipTableCell__right: 'tooltipTableCell__right'
  readonly tooltipTableCellDot: 'tooltipTableCellDot'
  readonly 'highcharts-loading': 'highcharts-loading'
  readonly 'highcharts-loading-inner': 'highcharts-loading-inner'
  readonly 'highcharts-tooltip-container': 'highcharts-tooltip-container'
  readonly 'highcharts-scrollable-mask': 'highcharts-scrollable-mask'
}
export = classNames
