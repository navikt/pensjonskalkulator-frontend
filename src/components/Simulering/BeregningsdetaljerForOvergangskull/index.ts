export { AlderspensjonDetaljerGrunnlag } from './AlderspensjonDetaljerGrunnlag'
export { AfpDetaljerGrunnlag } from './AfpDetaljerGrunnlag'
