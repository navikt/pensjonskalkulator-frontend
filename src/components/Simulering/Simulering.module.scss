@use '../../scss/variables';

.section {
  margin-bottom: var(--a-spacing-6);
}

.title {
  margin-bottom: var(--a-spacing-2);
}

.loader {
  display: inline-block;
  position: relative;
  width: 5rem;
  height: 5rem;

  & div {
    box-sizing: border-box;
    display: block;
    position: absolute;
    width: 4rem;
    height: 4rem;
    margin: var(--a-spacing-2);
    border: var(--a-spacing-2) solid var(--a-gray-600);
    border-radius: 50%;
    animation: lds-ring 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
    border-color: var(--a-gray-600) transparent transparent;

    &:nth-child(1) {
      animation-delay: -0.45s;
    }

    &:nth-child(2) {
      animation-delay: -0.3s;
    }

    &:nth-child(3) {
      animation-delay: -0.15s;
    }
  }
}

.tooltip {
  right: 0.375rem;
  background-color: var(--a-white);
  min-height: 115px;
  z-index: 3;
  position: relative;

  @media (min-width: variables.$a-breakpoint-lg) {
    min-height: 0;
    right: unset;
  }

  & path {
    filter: drop-shadow(0 0 5px var(--a-grayalpha-400));
  }

  & > span {
    position: relative !important;
    width: 100% !important;
    display: block;
    box-shadow: 0 0 var(--a-spacing-1) var(--a-grayalpha-400);
    border-radius: var(--a-spacing-1);
    padding: var(--a-spacing-3) var(--a-spacing-4);

    @media (min-width: variables.$a-breakpoint-lg) {
      background-color: var(--a-white);
      position: absolute !important;
      width: auto !important;
      box-shadow: none;
      padding: 0;
    }
  }

  &Table {
    position: relative;
    width: 100%;
    table-layout: fixed;
    border-collapse: separate;
    z-index: 4;

    @media (min-width: variables.$a-breakpoint-lg) {
      width: 22rem;
      border-width: var(--a-spacing-3) var(--a-spacing-4);
      border-style: solid;
      border-color: var(--a-white);
      border-radius: 0.25rem;
    }

    &HeaderCell {
      font-family: var(--a-font-family) !important;
      padding: 0.25rem 0;
      line-height: 1;

      &__left {
        width: 70%;
        white-space: break-spaces;
        font-weight: var(--a-font-weight-regular);
        font-size: var(--a-font-size-medium) !important;
        text-align: left;
        padding-right: var(--a-spacing-3);

        @media (width >= 400px) {
          width: auto;
          white-space: unset;
        }
      }

      &__right {
        font-size: var(--a-font-size-large) !important;
        font-weight: var(--a-font-weight-bold);
        text-align: right;
      }
    }

    &Cell {
      width: 70%;
      white-space: break-spaces;
      font-family: var(--a-font-family);
      font-size: var(--a-font-size-small);
      vertical-align: middle;
      padding: 0.188rem 0;
      line-height: 1;

      @media (width >= 400px) {
        width: auto;
        white-space: unset;
      }

      &__right {
        font-weight: var(--a-font-weight-bold);
        text-align: right;
      }

      &Dot {
        height: var(--a-spacing-3);
        width: var(--a-spacing-3);
        border-radius: 50%;
        display: inline-block;
        vertical-align: middle;
        margin-right: var(--a-spacing-1);
        margin-top: -2px;
      }
    }
  }

  &Line {
    position: absolute;
    width: 1px;
    border-right: 1px dashed var(--a-white);
    background-color: var(--a-grayalpha-500);

    @media (min-width: variables.$a-breakpoint-lg) {
      display: none;
    }
  }
}

:global {
  .highcharts-loading {
    opacity: 1 !important;
    background-color: rgb(255 255 255 / 70%) !important;
  }

  .highcharts-loading-inner {
    display: block;
    width: 80vw;

    @media (min-width: variables.$a-breakpoint-sm) {
      width: 100%;
    }
  }

  .highcharts-tooltip-container {
    left: 1.125rem !important;
    right: 1.125rem;

    @media (min-width: variables.$a-breakpoint-xl) {
      margin-left: auto;
      margin-right: auto;
      width: 100% !important;
      max-width: calc(variables.$a-breakpoint-xl-down - 2rem);
    }

    & > svg {
      display: none;
    }
  }

  .highcharts-scrolling-parent {
    // This creates a new stacking context
    isolation: isolate;

    // This removes the stacking context from that node
    .highcharts-container {
      z-index: auto !important;
    }

    /* y-axis */
    .highcharts-axis-labels.highcharts-yaxis-labels {
      z-index: 2;
      width: 30px;
      height: 100%;
      text-shadow:
        1px 1px 2px #fff,
        0 0 1em #fff,
        0 0 0.2em #fff;

      @media (min-width: variables.$a-breakpoint-lg) {
        width: 70px;
      }

      &::after {
        content: '';
        display: block;
        position: absolute;
        bottom: 0.25rem;
        height: 80%;
        width: 100%;
        z-index: -1;
        background: linear-gradient(
          33deg,
          rgb(255 255 255 / 75%) 40%,
          rgb(255 255 255 / 0%) 80%
        );

        @media (min-width: variables.$a-breakpoint-lg) {
          bottom: 4rem;
        }
      }
    }

    .highcharts-scrollable-mask {
      fill: white;
      filter: drop-shadow(-5px 0 3px rgb(255 255 255 / 50%));
    }
  }

  .highcharts-root path:hover {
    cursor: pointer;
  }

  .highcharts-background {
    fill: transparent;
  }

  .highcharts-legend {
    pointer-events: none;
  }

  .highcharts-legend-item:hover {
    color: var(--a-text-default) !important;
  }
}

@keyframes lds-ring {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}

::-webkit-full-page-media,
:future,
:root .tooltip path {
  stroke-width: 1px;
  stroke: var(--a-gray-500);
}

::-webkit-full-page-media,
:future,
:root .tooltipTable {
  @media (min-width: variables.$a-breakpoint-lg) {
    width: unset;
    border-width: 0;
  }
}

::-webkit-full-page-media,
:future,
:root .tooltipTableHeaderCell,
:root .tooltipTableCell {
  @media (min-width: variables.$a-breakpoint-lg) {
    padding: 0;
  }
}
