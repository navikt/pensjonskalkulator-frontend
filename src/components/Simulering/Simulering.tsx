import React from 'react'
import { FormattedMessage } from 'react-intl'

import { Heading, HeadingProps } from '@navikt/ds-react'
import Highcharts, { SeriesColumnOptions, XAxisOptions } from 'highcharts'
import HighchartsReact from 'highcharts-react-official'

import { TabellVisning } from '@/components/TabellVisning'
import {
  usePensjonsavtalerQuery,
  useOffentligTpQuery,
  useGetUtvidetSimuleringsresultatFeatureToggleQuery,
} from '@/state/api/apiSlice'
import {
  generateOffentligTpRequestBody,
  generatePensjonsavtalerRequestBody,
} from '@/state/api/utils'
import { useAppSelector } from '@/state/hooks'
import {
  selectCurrentSimulation,
  selectSamtykke,
  selectUfoeregrad,
  selectSivilstand,
  selectSamboer,
  selectAfp,
  selectIsEndring,
  selectFoedselsdato,
} from '@/state/userInput/selectors'

import { useSimuleringChartLocalState } from './hooks'
import { SimuleringEndringBanner } from './SimuleringEndringBanner/SimuleringEndringBanner'
import { SimuleringGrafNavigation } from './SimuleringGrafNavigation/SimuleringGrafNavigation'
import { SimuleringPensjonsavtalerAlert } from './SimuleringPensjonsavtalerAlert/SimuleringPensjonsavtalerAlert'
import { Simuleringsdetaljer } from './Simuleringsdetaljer/Simuleringsdetaljer'

import styles from './Simulering.module.scss'

export function Simulering(props: {
  isLoading: boolean
  headingLevel: HeadingProps['level']
  aarligInntektFoerUttakBeloep: string
  alderspensjonListe?: AlderspensjonPensjonsberegning[]
  afpPrivatListe?: AfpPrivatPensjonsberegning[]
  afpOffentligListe?: AfpPrivatPensjonsberegning[]
  alderspensjonMaanedligVedEndring?: AlderspensjonMaanedligVedEndring
  showButtonsAndTable?: boolean
  detaljer?: {
    trygdetid?: number
    opptjeningsgrunnlag?: SimulertOpptjeningGrunnlag[]
  }
}) {
  const {
    isLoading,
    headingLevel,
    aarligInntektFoerUttakBeloep,
    alderspensjonListe,
    afpPrivatListe,
    afpOffentligListe,
    alderspensjonMaanedligVedEndring,
    showButtonsAndTable,
    detaljer,
  } = props

  const harSamtykket = useAppSelector(selectSamtykke)
  const ufoeregrad = useAppSelector(selectUfoeregrad)
  const afp = useAppSelector(selectAfp)
  const sivilstand = useAppSelector(selectSivilstand)
  const foedselsdato = useAppSelector(selectFoedselsdato)
  const isEndring = useAppSelector(selectIsEndring)
  const harSamboer = useAppSelector(selectSamboer)
  const {
    uttaksalder,
    aarligInntektVsaHelPensjon,
    gradertUttaksperiode,
    utenlandsperioder,
  } = useAppSelector(selectCurrentSimulation)
  const { data: utvidetSimuleringsresultatFeatureToggle } =
    useGetUtvidetSimuleringsresultatFeatureToggleQuery()

  const chartRef = React.useRef<HighchartsReact.RefObject>(null)

  const [offentligTpRequestBody, setOffentligTpRequestBody] = React.useState<
    OffentligTpRequestBody | undefined
  >(undefined)

  const [pensjonsavtalerRequestBody, setPensjonsavtalerRequestBody] =
    React.useState<PensjonsavtalerRequestBody | undefined>(undefined)

  const {
    data: offentligTpData,
    isFetching: isOffentligTpLoading,
    isError: isOffentligTpError,
  } = useOffentligTpQuery(offentligTpRequestBody as OffentligTpRequestBody, {
    skip: !offentligTpRequestBody || !harSamtykket || !uttaksalder,
  })

  const {
    data: pensjonsavtalerData,
    isFetching: isPensjonsavtalerLoading,
    isSuccess: isPensjonsavtalerSuccess,
    isError: isPensjonsavtalerError,
  } = usePensjonsavtalerQuery(
    pensjonsavtalerRequestBody as PensjonsavtalerRequestBody,
    {
      skip: !pensjonsavtalerRequestBody || !harSamtykket || !uttaksalder,
    }
  )

  React.useEffect(() => {
    if (harSamtykket && uttaksalder) {
      setOffentligTpRequestBody(
        generateOffentligTpRequestBody({
          afp,
          foedselsdato,
          harSamboer,
          aarligInntektFoerUttakBeloep: aarligInntektFoerUttakBeloep ?? '0',
          gradertUttak: gradertUttaksperiode ? gradertUttaksperiode : undefined,
          heltUttak: {
            uttaksalder,
            aarligInntektVsaPensjon: aarligInntektVsaHelPensjon,
          },
          utenlandsperioder,
        })
      )
      setPensjonsavtalerRequestBody(
        generatePensjonsavtalerRequestBody({
          ufoeregrad,
          afp,
          sivilstand,
          harSamboer,
          aarligInntektFoerUttakBeloep: aarligInntektFoerUttakBeloep ?? '0',
          gradertUttak: gradertUttaksperiode ? gradertUttaksperiode : undefined,
          heltUttak: {
            uttaksalder,
            aarligInntektVsaPensjon: aarligInntektVsaHelPensjon,
          },
        })
      )
    }
  }, [harSamtykket, uttaksalder])

  const [
    chartOptions,
    showVisFaerreAarButton,
    showVisFlereAarButton,
    isPensjonsavtaleFlagVisible,
  ] = useSimuleringChartLocalState({
    styles,
    chartRef,
    foedselsdato,
    isEndring,
    uttaksalder,
    gradertUttaksperiode,
    aarligInntektFoerUttakBeloep,
    aarligInntektVsaHelPensjon,
    isLoading,
    alderspensjonListe,
    afpPrivatListe,
    afpOffentligListe,
    pensjonsavtaler: {
      isLoading: isPensjonsavtalerLoading,
      data: pensjonsavtalerData,
    },
    offentligTp: {
      isLoading: isOffentligTpLoading,
      data: offentligTpData,
    },
  })

  return (
    <section className={styles.section}>
      <Heading level={headingLevel} size="medium" className={styles.title}>
        <FormattedMessage id="beregning.highcharts.title" />
      </Heading>

      <SimuleringEndringBanner
        heltUttaksalder={uttaksalder}
        gradertUttaksperiode={gradertUttaksperiode ?? undefined}
        alderspensjonMaanedligVedEndring={alderspensjonMaanedligVedEndring}
      />

      <div role="img" aria-labelledby="alt-chart-title">
        <div id="alt-chart-title" hidden>
          <FormattedMessage id="beregning.alt_tekst" />
        </div>
        <div data-testid="highcharts-aria-wrapper" aria-hidden={true}>
          <HighchartsReact
            ref={chartRef}
            highcharts={Highcharts}
            options={chartOptions}
          />
        </div>
      </div>
      {showButtonsAndTable && (
        <SimuleringGrafNavigation
          showVisFaerreAarButton={showVisFaerreAarButton}
          showVisFlereAarButton={showVisFlereAarButton}
        />
      )}
      <SimuleringPensjonsavtalerAlert
        isPensjonsavtaleFlagVisible={isPensjonsavtaleFlagVisible}
        pensjonsavtaler={{
          isLoading: isPensjonsavtalerLoading,
          isError: isPensjonsavtalerError,
          isSuccess: isPensjonsavtalerSuccess,
          data: pensjonsavtalerData,
        }}
        offentligTp={{
          isError: isOffentligTpError,
          data: offentligTpData,
        }}
      />

      {showButtonsAndTable && (
        <TabellVisning
          series={chartOptions.series as SeriesColumnOptions[]}
          aarArray={(chartOptions?.xAxis as XAxisOptions).categories}
        />
      )}
      {/* c8 ignore next 6 - detaljer skal kun vises i dev for test form√•l */}
      {utvidetSimuleringsresultatFeatureToggle?.enabled && detaljer && (
        <Simuleringsdetaljer
          alderspensjonListe={alderspensjonListe}
          detaljer={detaljer}
        />
      )}
    </section>
  )
}
