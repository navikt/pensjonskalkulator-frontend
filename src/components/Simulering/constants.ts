export const MAX_UTTAKSALDER = 77
export const COLUMN_WIDTH = 25
export const TOOLTIP_YPOS = 35

export const SERIES_COLORS = {
  SERIE_COLOR_INNTEKT: 'var(--a-gray-500)',
  SERIE_COLOR_AFP: 'var(--a-purple-400)',
  SERIE_COLOR_TP: 'var(--a-green-400)',
  SERIE_COLOR_ALDERSPENSJON: 'var(--a-deepblue-500)',
  SERIE_COLOR_FADED_INNTEKT: 'var(--a-gray-300)',
  SERIE_COLOR_FADED_AFP: 'var(--a-purple-200)',
  SERIE_COLOR_FADED_TP: 'var(--a-green-200)',
  SERIE_COLOR_FADED_ALDERSPENSJON: 'var(--a-deepblue-200)',
}

export const SERIES_DEFAULT = {
  SERIE_INNTEKT: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Pensjonsgivende inntekt',
    color: SERIES_COLORS.SERIE_COLOR_INNTEKT,
  },
  SERIE_TP: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Pensjonsavtaler (arbeidsgivere m.m.)',
    color: SERIES_COLORS.SERIE_COLOR_TP,
  },
  SERIE_AFP: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'AFP (Avtalefestet pensjon)',
    color: SERIES_COLORS.SERIE_COLOR_AFP,
  },
  SERIE_ALDERSPENSJON: {
    type: 'column',
    pointWidth: COLUMN_WIDTH,
    name: 'Alderspensjon (NAV)',
    color: SERIES_COLORS.SERIE_COLOR_ALDERSPENSJON,
  },
}

export const highchartsScrollingSelector = '.highcharts-scrolling'
