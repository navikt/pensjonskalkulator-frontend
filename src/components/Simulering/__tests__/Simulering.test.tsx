import { describe, it, vi } from 'vitest'

import alderspensjonData from '../../../mocks/data/alderspensjon/67.json' assert { type: 'json' }
import { Simulering } from '../Simulering'
import { mockResponse } from '@/mocks/server'
import * as apiSliceUtils from '@/state/api/apiSlice'
import {
  userInputInitialState,
  Simulation,
} from '@/state/userInput/userInputReducer'
import { act, render, screen, waitFor } from '@/test-utils'

describe('Simulering', () => {
  const currentSimulation: Simulation = {
    startAlder: 70,
    startMaaned: 5,
    uttaksgrad: 100,
    aarligInntekt: 0,
  }

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('Når brukeren har samtykket, henter og viser Pensjonsavtaler', async () => {
    const usePensjonsavtalerQueryMock = vi.spyOn(
      apiSliceUtils,
      'usePensjonsavtalerQuery'
    )
    const { container } = render(
      <Simulering
        alderspensjon={alderspensjonData}
        showAfp={false}
        showButtonsAndTable={false}
      />,
      {
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            samtykke: true,
            afp: 'nei',
            currentSimulation: { ...currentSimulation },
          },
        },
      }
    )
    await waitFor(async () => {
      expect(await screen.findByTestId('highcharts-done-drawing')).toBeVisible()
    })
    expect(usePensjonsavtalerQueryMock).toHaveBeenLastCalledWith(
      {
        antallInntektsaarEtterUttak: 0,
        uttaksperioder: [
          {
            aarligInntekt: 0,
            grad: 100,
            startAlder: 70,
            startMaaned: 5,
          },
        ],
      },
      { skip: false }
    )
    // Nødvendig for at animasjonen rekker å bli ferdig
    await act(async () => {
      await new Promise((r) => setTimeout(r, 500))
    })

    expect(
      container.getElementsByClassName('highcharts-container')
    ).toHaveLength(1)
    const legendContainer =
      container.getElementsByClassName('highcharts-legend')
    const legendItems = (
      legendContainer[0] as HTMLElement
    ).getElementsByClassName('highcharts-legend-item')
    expect(legendItems).toHaveLength(3)
  })

  it('Når brukeren har samtykket og valgt AFP-privat, henter og viser AFP og Pensjonsavtaler når brukeren har valgt', async () => {
    const { container } = render(
      <Simulering
        alderspensjon={alderspensjonData}
        showAfp={true}
        showButtonsAndTable={true}
      />,
      {
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            samtykke: true,
            afp: 'ja_privat',
            currentSimulation: { ...currentSimulation },
          },
        },
      }
    )

    await waitFor(async () => {
      expect(await screen.findByTestId('highcharts-done-drawing')).toBeVisible()
    })
    // Nødvendig for at animasjonen rekker å bli ferdig
    await act(async () => {
      await new Promise((r) => setTimeout(r, 500))
    })

    expect(
      container.getElementsByClassName('highcharts-container')
    ).toHaveLength(1)
    const legendContainer =
      container.getElementsByClassName('highcharts-legend')
    const legendItems = (
      legendContainer[0] as HTMLElement
    ).getElementsByClassName('highcharts-legend-item')
    expect(legendItems).toHaveLength(4)
  })

  it('Når brukeren ikke samtykker og ikke velger AFP, viser kun alderspensjon og inntekt', async () => {
    const usePensjonsavtalerQueryMock = vi.spyOn(
      apiSliceUtils,
      'usePensjonsavtalerQuery'
    )
    const { container } = render(
      <Simulering
        alderspensjon={alderspensjonData}
        showAfp={false}
        showButtonsAndTable={false}
      />,
      {
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            samtykke: false,
            currentSimulation: { ...currentSimulation },
          },
        },
      }
    )

    await waitFor(async () => {
      expect(await screen.findByTestId('highcharts-done-drawing')).toBeVisible()
    })
    expect(usePensjonsavtalerQueryMock.mock.calls[0][1]).toEqual({
      skip: true,
    })
    // Nødvendig for at animasjonen rekker å bli ferdig
    await act(async () => {
      await new Promise((r) => setTimeout(r, 500))
    })

    expect(
      container.getElementsByClassName('highcharts-container')
    ).toHaveLength(1)
    expect(
      container.getElementsByClassName('highcharts-legend-item')
    ).toHaveLength(4)
  })

  it('Når brukeren har samtykket og har valgt AFP-privat, viser alderspensjon, inntekt og AFP', async () => {
    const { container } = render(
      <Simulering
        alderspensjon={alderspensjonData}
        showAfp={true}
        showButtonsAndTable={true}
      />,
      {
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            samtykke: false,
            afp: 'ja_privat',
            currentSimulation: { ...currentSimulation },
          },
        },
      }
    )

    await waitFor(async () => {
      expect(await screen.findByTestId('highcharts-done-drawing')).toBeVisible()
    })
    // Nødvendig for at animasjonen rekker å bli ferdig
    await act(async () => {
      await new Promise((r) => setTimeout(r, 500))
    })

    expect(
      container.getElementsByClassName('highcharts-container')
    ).toHaveLength(1)
    expect(
      container.getElementsByClassName('highcharts-legend-item')
    ).toHaveLength(6)
  })

  it('viser infomelding om pensjonsavtaler når brukeren har en pensjonsavtale som begynner før uttaksalderen', async () => {
    mockResponse('/pensjonsavtaler', {
      status: 200,
      json: {
        avtaler: [
          {
            produktbetegnelse: 'Storebrand',
            kategori: 'PRIVAT_TJENESTEPENSJON',
            startAlder: 62,
            sluttAlder: 72,
            utbetalingsperioder: [
              {
                startAlder: 62,
                startMaaned: 1,
                sluttAlder: 72,
                sluttMaaned: 1,
                aarligUtbetaling: 31298,
                grad: 100,
              },
            ],
          },
        ],
        utilgjengeligeSelskap: [],
      },
      method: 'post',
    })
    render(
      <Simulering
        alderspensjon={alderspensjonData}
        showAfp={false}
        showButtonsAndTable={true}
      />,
      {
        preloadedState: {
          userInput: {
            ...userInputInitialState,
            samtykke: true,
            currentSimulation: { ...currentSimulation },
          },
        },
      }
    )
    await waitFor(async () => {
      expect(await screen.findByText('Beregning')).toBeVisible()
      expect(
        await screen.findByText('beregning.pensjonsavtaler.info')
      ).toBeVisible()
    })
  })

  it('viser tabell', async () => {
    render(<Simulering showAfp={true} showButtonsAndTable={true} />, {
      preloadedState: {
        userInput: {
          ...userInputInitialState,
          samtykke: true,
          currentSimulation: { ...currentSimulation },
        },
      },
    })

    expect(screen.getByText('Vis tabell av beregningen')).toBeVisible()
  })

  // TODO PEK-97
  // it('henter pensjonsavtaler og viser riktig feilmelding ved feil', async () => {
  //   mockErrorResponse('/pensjonsavtaler', {
  //     status: 500,
  //     json: "Beep boop I'm an error!",
  //     method: 'post',
  //   })
  //   render(<Simulering showAfp={false} showButtonsAndTable={false} />, {
  //     preloadedState: {
  //       userInput: { ...userInputInitialState, samtykke: true },
  //     },
  //   })

  //   await waitFor(async () => {
  //     expect(
  //       await screen.findByText(
  //         'Vi klarte ikke å hente pensjonsavtalene dine fra Norsk Pensjon. Prøv igjen senere.'
  //       )
  //     ).toBeVisible()
  //   })
  // })
})
