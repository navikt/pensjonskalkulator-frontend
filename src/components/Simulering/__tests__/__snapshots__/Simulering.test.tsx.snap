// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Simulering > Gitt at brukeren IKKE samtykker > Når brukeren ikke velger AFP, viser kun inntekt og alderspensjon 1`] = `
[
  <g
    class="highcharts-legend-item highcharts-column-series highcharts-color-undefined highcharts-series-0"
    data-z-index="1"
    transform="translate(0,-5)"
  >
    <rect
      class="highcharts-point"
      data-z-index="3"
      fill="var(--a-gray-500)"
      height="14"
      rx="7"
      ry="7"
      width="14"
      x="1"
      y="6"
    />
  </g>,
  <g
    class="highcharts-legend-item highcharts-column-series highcharts-color-undefined highcharts-series-1"
    data-z-index="1"
    transform="translate(33,-5)"
  >
    <rect
      class="highcharts-point"
      data-z-index="3"
      fill="var(--a-deepblue-500)"
      height="14"
      rx="7"
      ry="7"
      width="14"
      x="1"
      y="6"
    />
  </g>,
]
`;

exports[`Simulering > Gitt at brukeren IKKE samtykker > Når brukeren velger AFP-privat, viser inntekt, alderspensjon og AFP 1`] = `
[
  <g
    class="highcharts-legend-item highcharts-column-series highcharts-color-undefined highcharts-series-0"
    data-z-index="1"
    transform="translate(0,-5)"
  >
    <rect
      class="highcharts-point"
      data-z-index="3"
      fill="var(--a-gray-500)"
      height="14"
      rx="7"
      ry="7"
      width="14"
      x="1"
      y="6"
    />
  </g>,
  <g
    class="highcharts-legend-item highcharts-column-series highcharts-color-undefined highcharts-series-1"
    data-z-index="1"
    transform="translate(33,-5)"
  >
    <rect
      class="highcharts-point"
      data-z-index="3"
      fill="var(--a-purple-400)"
      height="14"
      rx="7"
      ry="7"
      width="14"
      x="1"
      y="6"
    />
  </g>,
  <g
    class="highcharts-legend-item highcharts-column-series highcharts-color-undefined highcharts-series-2"
    data-z-index="1"
    transform="translate(66,-5)"
  >
    <rect
      class="highcharts-point"
      data-z-index="3"
      fill="var(--a-deepblue-500)"
      height="14"
      rx="7"
      ry="7"
      width="14"
      x="1"
      y="6"
    />
  </g>,
]
`;
